allennlp/models/semantic_parsing/wikitables/wikitables_semantic_parser.py

```diff
@@ -518,133 +507,154 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        """
+        This method creates the GrammarState object that's used for decoding.  Part of creating
+        that is creating the `valid_actions` dictionary, which contains embedded representations of
+        all of the valid actions.  So, we create that here as well.
+
+        The inputs to this method are for a `single instance in the batch`; none of the tensors we
+        create here are batched.  We grab the global action ids from the input
+        ``ProductionRuleArrays``, and we use those to embed the valid actions for every
+        non-terminal type.  We use the input ``linking_scores`` for non-global actions.
+
+        Parameters
+        ----------
+        world : ``WikiTablesWorld``
+            From the input to ``forward`` for a single batch instance.
+        possible_actions : ``List[ProductionRuleArray]``
+            From the input to ``forward`` for a single batch instance.
+        linking_scores : ``torch.Tensor``
+            Assumed to have shape ``(num_entities, num_question_tokens)`` (i.e., there is no batch
+            dimension).
+        entity_types : ``torch.Tensor``
+            Assumed to have shape ``(num_entities,)`` (i.e., there is no batch dimension).
+        """
+        action_map = {}
+        for action_index, action in enumerate(possible_actions):
             action_string = action[0]
-            action_mapping[action_string] = i
-        translated_valid_actions = {}
+            action_map[action_string] = action_index
+        entity_map = {}
+        for entity_index, entity in enumerate(world.table_graph.entities):
+            entity_map[entity] = entity_index
+
+        valid_actions = world.get_valid_actions()
+        translated_valid_actions: Dict[str, Dict[str, Tuple[torch.Tensor, torch.Tensor, List[int]]]] = {}
         for key, action_strings in valid_actions.items():
-            translated_valid_actions[key] = [action_mapping[action_string]
-                                             for action_string in action_strings]
+            translated_valid_actions[key] = {}
+            # `key` here is a non-terminal from the grammar, and `action_strings` are all the valid
+            # productions of that non-terminal.  We'll first split those productions by global vs.
+            # linked action.
+            action_indices = [action_map[action_string] for action_string in action_strings]
+            production_rule_arrays = [(possible_actions[index], index) for index in action_indices]
+            global_actions = []
+            linked_actions = []
+            for production_rule_array, action_index in production_rule_arrays:
+                if production_rule_array[1]:
+                    global_actions.append((production_rule_array[2], action_index))
+                else:
+                    linked_actions.append((production_rule_array[0], action_index))
+
+            # Then we get the ebmedded representations of the global actions.
```

Comment:
embedded

Reply:
I thought I fixed that already...  Thanks.  Done.